### PR TITLE
Adding proper accomodation for textScaleFactor in bottom navigation bar.

### PR DESCRIPTION
**Pull request by gspencergoog**
The first PR of many trying to address the behavior of fixed-sized widgets in the presence of text scale factor changes.

### CODE REVIEW

**Comment by gspencergoog** [comment]
Also, changed the height of the bottom nav bar, shrinking it by 4 pixels to match the [Material spec](https://material.io/guidelines/components/bottom-navigation.html) at 56 dp.

File: packages/flutter/lib/src/material/bottom_navigation_bar.dart [hunk 1]
```diff
@@ -412,7 +415,10 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
           )
         ),
         new SizedBox(
-          height: kBottomNavigationBarHeight,
+          height: math.max(
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/bottom_navigation_bar.dart]: Can you add some tests verifying that this does the right thing?

Comment (gspencergoog) [reply packages/flutter/lib/src/material/bottom_navigation_bar.dart]: Duh.  Yes, sorry, just got over excited about getting it in.

**Comment by gspencergoog** [comment]
How much should I worry about people putting non-text widgets in for the labels?  Right now I assume that there's a text widget there.

File: packages/flutter/lib/src/material/bottom_navigation_bar.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/material/bottom_navigation_bar.dart]: I know that feeling :)

Comment (gspencergoog) [reply packages/flutter/lib/src/material/bottom_navigation_bar.dart]: Test added.

**Comment by Hixie** [comment]
You should make no assumptions about what people will put into child slots.

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart [hunk 2]
```diff
@@ -95,31 +95,31 @@ class _BottomNavigationDemoState extends State<BottomNavigationDemo>
     _navigationViews = <NavigationIconView>[
       new NavigationIconView(
         icon: const Icon(Icons.access_alarm),
-        title: const Text('Alarm'),
+        title: const Text('Alarm', overflow: TextOverflow.ellipsis),
```
Comment (Hixie) [inline examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: We should make the bottom navigation widget do this by default (by setting it on the DefaultTextStyle around its children), the user shouldn't have to explicitly set this.

File: packages/flutter/lib/src/material/bottom_navigation_bar.dart [hunk 3]
```diff
@@ -412,7 +415,10 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
...
+            kBottomNavigationBarHeight,
+            widget.iconSize + _bottomMargin + _topMargin + _fontSize * MediaQuery.of(context).textScaleFactor
```
Comment (Hixie) [inline packages/flutter/lib/src/material/bottom_navigation_bar.dart]: nit: trailing comma

File: packages/flutter/lib/src/material/bottom_navigation_bar.dart [hunk 4]
```diff
@@ -375,12 +378,12 @@ class _BottomNavigationBarState extends State<BottomNavigationBar> with TickerPr
...
                       child: new Container(
-                        margin: const EdgeInsets.only(bottom: 10.0),
+                        margin: const EdgeInsets.only(bottom: _bottomMargin),
                         child: new FadeTransition(
                           opacity: _animations[i],
                           child: DefaultTextStyle.merge(
                             style: const TextStyle(
-                              fontSize: 14.0,
+                              fontSize: _fontSize,
                               color: Colors.white
```
Comment (Hixie) [inline packages/flutter/lib/src/material/bottom_navigation_bar.dart]: while you're here: nit: trailing comma

**Comment by gspencergoog** [comment]
OK, then how do I detect that a child slot contains at least one Text widget, and what the effective scale is on the contained text?  After all, they could put a MediaQuery with its own text scale factor into a child slot.

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart (hunk 2 above)
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: OK, Done.

File: packages/flutter/lib/src/material/bottom_navigation_bar.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/bottom_navigation_bar.dart]: Done

**Comment by Hixie** [comment]
> how do I detect that a child slot contains at least one Text widget

We should be agnostic as to what is in the child slots. For example, the user could have invented their own Text-analogue widget that we don't know about, and it should work the same way.

**Comment by gspencergoog** [comment]
Yes, ideally, we'd be able to measure the size of the child, whether it contains text or not, and a Text child would just scale itself using the textScaleFactor and report the correct size.  Unfortunately, that would mean more than one pass for layout, right?  I don't thing we should add a whole new pass to the layout just so we can be agnostic of whether or not a text child is inserted in a place where text is almost always inserted. Are there any techniques in the framework for determining child dimensions as part of the build pass?  I.e. Can we determine "natural" size for a child at build time?

**Comment by gspencergoog** [comment]
Thinking more about this, if we want **any** widget to be able to be placed into the child, then I need to rewrite this entire widget to use rows, columns, etc. and to wrap it in a min height constraint, rather than just calculating what the height should be and assuming a text child.

Hopefully this isn't par for the course, or I'll be doing a lot of rewriting of widgets.. :-)

**Comment by Hixie** [comment]
I think to answer your questions we have to first decide on the following:
- If the child widget wants to be as tall as possible (e.g. a Center widget), how tall should the parent be?
- If the child widget wants to be 1000px tall if possible (e.g. an Image), how tall should the parent be?
- If the child widget wants to be as tall as possible unless that's infinite in which case it'll settle for 400px (e.g. a Placeholder), how tall should the parent be?
- If the child widget wants to be 20px tall (e.g. big Text), how tall should the parent be?
- If the child widget wants to be as tall as possible unless that's infinite in which case it'll become 0px tall (some widgets do this), how tall should the parent be?

The answers could vary:
- Always 56px
- Always 56px * the font size scale, if the scale is greater than 1.0
- Whatever the child wants to be
- Whatever the child wants to be, up to some maximum

**Comment by gspencergoog** [comment]
The Material Design team's response when I asked about text scaling was basically: if there's text in a container, then the text should scale up with the textScaleFactor until the point where it fills up any available space, and then the container should grow to contain the scaled text so it isn't clipped vertically.  That's different than the answers you mentioned, but it's closest to "Whatever the child wants to be", since theoretically the text scale could be huge.

Considering the use case, at least for bottom navigation, I would say that since the bottom nav is [described in the spec](https://material.io/guidelines/components/bottom-navigation.html#bottom-navigation-usage) specifically as having icons with text below them (there's an entire section called "Icons and text"), and with a specific size in dp, we should assume that is how the user is expected to use it.  If they do otherwise, then they are off the reservation, and they get what they get.  Worst case, they get some odd margins when they put something in there that doesn't respect textScaleFactor, and the user has set large/small text in their system settings. In any case, if the textScaleFactor is 1.0, then given *any* child, the bottom nav bar should be exactly 56dp high, so they can't really put huge things in there anyhow.

Incidentally, we also don't implement all of the spec: 

- When there are only three actions, display both icons and text labels at all times
- If there are four or five actions, display inactive views as icons only

**Comment by gspencergoog** [comment]
I spoke with Ian offline, and we decided that this widget needs to be rewritten to take arbitrarily sized children for the labels.  Closing this PR: I'll open a new one when I've rewritten it.

**Pull request closed without merging**